dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs

```diff
@@ -97,7 +95,31 @@ internal KernelFunction(string name, string description, IReadOnlyList<KernelPar
             Parameters = parameters,
             ReturnParameter = returnParameter ?? KernelReturnParameterMetadata.Empty,
         };
-        this.ExecutionSettings = executionSettings;
+        this._executionSettings = executionSettings;
+    }
+
+    /// <summary>
+    /// Gets a copy of the prompt execution settings associated with this function.
+    /// </summary>
+    /// <remarks>
+    /// The returned dictionary is a deep copy of the settings associated with this function, and
+    /// the <see cref="PromptExecutionSettings"/> can be changed without affecting the function's
+    /// copy of the settings.
+    /// </remarks>
+    public IReadOnlyDictionary<string, PromptExecutionSettings>? GetExecutionSettings()
+    {
+        if (this._executionSettings == null)
+        {
+            return null;
+        }
+
+        var dictionary = new Dictionary<string, PromptExecutionSettings>();
+        foreach (var keyValue in this._executionSettings)
+        {
+            dictionary.Add(keyValue.Key, (PromptExecutionSettings)keyValue.Value.Clone());
+        }
```

Comment:
I'm concerned about the expense of this.  How common will it be for a) functions to have non-null _executionSettings and b) for this GetExecutionSettings() to be accessed.  Is it called every time a function is invoked? That makes me nervous.

Is there an alternative design where something truly readonly could be surfaced?

Or maybe we give up on the guaranteed immutability and just document "don't mutate this"?

Reply:
What information from the PromptExecutionSettings is actually required by consumers?

Reply:
Typically for a service selector implementation the model id and service id would be used. But I could also imagine a scenario where someone wants to switch based on the max tokens that are configured.

The obvious risk is some code mutating the execution settings during  a request and impacting all future uses of that function.

I'll look at lazily creating a read only version of the execution settings for each instance and using that.

Reply:
What if we added a virtual Freeze method, where after calling it attempting to set any properties throw? And we'd freeze it as part of function construction.